code/espurna/mcp23s17.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+
+MCP23S17 MODULE
+
+Copyright (C) 2020 by Eddi De Pieri <eddi at depieri dot com>
+Copyright (C) 2016-2017 by Xose Pérez <xose dot perez at gmail dot com>
+Copyright (C) 2016 Plamen Kovandjiev <p.kovandiev@kmpelectronics.eu> & Dimitar Antonov <d.antonov@kmpelectronics.eu>
+
+*/
+
+#include "mcp23s17.h"
+
+#if MCP23S17_SUPPORT
+
+#define CS 0x0F
+
+#define READ_CMD  0x41
+#define WRITE_CMD 0x40 
+
+#define IODIR   0x00
+#define IPOL    0x01
+#define GPINTEN 0x02
+#define DEFVAL  0x03
+#define INTCON  0x04
+#define IOCON   0x05
+#define GPPU    0x06
+#define INTF    0x07
+#define INTCAP  0x08
+#define GPIO    0x09
+#define OLAT    0x0A 
+
+/**
+ * @brief Relay pins.
+ */
+const uint8_t RELAY_PINS[DUMMY_RELAY_COUNT] =
+{ REL1PIN, REL2PIN, REL3PIN, REL4PIN };
+
+/**
+ * @brief Input pins.
+ */
+const int OPTOIN_PINS[OPTOIN_COUNT] =
+{ IN1PIN, IN2PIN, IN3PIN, IN4PIN };
```

Comment:
I had noticed these are defined in hardware header.
- please prefix with module name
- can we provide default values?

Reply:
_"please prefix with module name"_ 
done

_can we provide default values?_ 
no idea on how to do that in the correct way

Reply:
'guards' I was mentioning earlier are constructs like:
```cpp
#ifndef FOO
#define FOO "bar"
#endif
```

`#if SOMETHING_SUPPORT` can be used to exclude code, it will always be 0 even without `#ifndef` block above. `#ifdef SOMETHING_SUPPORT`, however, means that we used either `-DSOMETHING_SUPPORT=0` or have a `#define SOMETHING_SUPPORT` (even without any value attached to it)